File: mcpobs/payload.py
"""Tool request and response capture.

WHY IT IS OFF BY DEFAULT, UNLIKE ERROR DETAIL
    Error detail (D49) is narrow: failing calls only, and the text is usually an
    exception message. This is every argument and every result of every call --
    a different magnitude in both volume and sensitivity. V2 §15 specifies
    payloads as "opt-in, redactable, size-limited, and separately retained", and
    a tool named `get_customer` returning a customer record is the normal case,
    not the edge case.

    So: one flag, easy to find, off until asked for.

        instrument(mcp, capture_payloads=True)

WHAT THE SAFEGUARDS ACTUALLY DO
    Redaction is pattern-based and therefore incomplete. It catches the obvious
    shapes -- a field called `password`, a bearer token, an API key -- and will
    miss a secret in a field called `note`. It reduces harm; it does not make
    capture safe. Anyone turning this on should know that, so the docstring says
    it rather than implying a guarantee.
"""

from __future__ import annotations

import json
import re
from typing import Any, Final

MAX_CHARS: Final = 2048
"""Truncation limit. Full results can be megabytes; a preview is for reading."""

REDACTED: Final = "[redacted]"

#: Field names whose VALUE is replaced wholesale. Matched case-insensitively as
#: a substring, so `api_key`, `apiKey` and `X-API-KEY` all hit.
SENSITIVE_KEYS: Final[tuple[str, ...]] = (
    "password", "passwd", "secret", "token", "api_key", "apikey",
    "authorization", "auth", "credential", "private_key", "session",
    "cookie", "signature", "access_key",
    # `requestState` carries RECORDED ELICITATION OUTCOMES -- the user's actual
    # answers (D28). It is hashed for MRTR correlation precisely because it must
    # not be stored; now that the whole params object is captured it would
    # otherwise ride along in plain text.
    "requeststate",
)

#: Keys that must NEVER be redacted, checked BEFORE the sensitive list.
#:
#: Pattern redaction is documented as incomplete (D56), but that was only ever
#: stated in one direction -- that it might MISS a secret. It also OVER-matches,
#: and that failure is quieter: `progressToken` contains "token", so the field
#: that correlates a progress notification to its request was being destroyed in
#: every captured payload. A missed secret is a risk you can reason about; a
#: silently deleted protocol field looks like the server never sent it.
#:
#: Deliberately an explicit list of PROTOCOL field names rather than a cleverer
#: matcher. Segment-aware matching would fix `author` (contains "auth") and not
#: `progressToken`, because "token" genuinely is a whole segment there -- the
#: distinction is what the field MEANS, which no string rule recovers.
#:
#: Known remaining false positives, stated rather than hidden: any key containing
#: `auth` (`author`), `session` (`sessionCount`) or `token` that is not listed
#: here is still redacted.
NEVER_REDACT: Final[frozenset[str]] = frozenset({
    "progresstoken",   # correlates progress notifications to their request
})

#: Value shapes that are secrets regardless of the field they sit in.
SENSITIVE_VALUES: Final[tuple[re.Pattern[str], ...]] = (
    re.compile(r"\bBearer\s+[A-Za-z0-9._~+/=-]{8,}", re.I),
    re.compile(r"\bsk-[A-Za-z0-9]{16,}"),                  # OpenAI-style
    re.compile(r"\bgh[pousr]_[A-Za-z0-9]{16,}"),           # GitHub
    re.compile(r"\bAKIA[0-9A-Z]{16}\b"),                   # AWS access key id
    re.compile(r"\beyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\."),  # JWT
)


class PayloadCapture:
    """Renders tool arguments and results into bounded, redacted previews."""

    def __init__(self, max_chars: int = MAX_CHARS, redact: bool = True) -> None:
        self.max_chars = max_chars
        self.redact = redact

    # -- public ------------------------------------------------------------
    def request(self, method: str, request_id: Any, params: Any) -> tuple[str, int]:
        """The JSON-RPC request, as (preview, original size).

        THE ACTUAL WIRE MESSAGE, not just `arguments`. This is an MCP
        observability product: the useful artefact is the protocol message you
        can compare against the spec or paste into a bug report, and the first
        version threw away everything except the arguments.

        What that discarded, all of it debugging-relevant:
          * `_meta.io.modelcontextprotocol/clientInfo` -- WHICH CLIENT called.
            The SDK sets no client attribute on the span, so this is the only
            place client identity appears at all, and V2 6.1 asks for exactly
            that ("which clients are calling which tools").
          * `_meta` clientCapabilities -- what the client claimed to support,
            the first thing to check on a capability error.
          * `_meta` protocolVersion and traceparent.
          * `name`, so a prompt or resource call is self-describing.
        """
        envelope = {
            "jsonrpc": "2.0",
            "id": request_id,
            "method": method,
            "params": params if isinstance(params, dict) else {},
        }
        return self._render(envelope)

    def response(self, request_id: Any, result: Any) -> tuple[str, int]:
        """The JSON-RPC response, as (preview, original size).

        Also the whole message, so `resultType`, `isError` and especially
        `structuredContent` -- the typed result, previously invisible -- survive.

        Works for EVERY method. The first version looked for `content` blocks
        and therefore returned nothing for prompts/get (which returns
        `messages`) or resources/read (which returns `contents`).
        """
        body = result if isinstance(result, dict) else self._as_dict(result)
        if not body:
            return "", 0
        return self._render({"jsonrpc": "2.0", "id": request_id, "result": body})

    @staticmethod
    def _as_dict(result: Any) -> dict[str, Any]:
        """Best-effort for a model rather than the sealed wire form."""
        for attr in ("model_dump", "dict"):
            dumper = getattr(result, attr, None)
            if callable(dumper):
                try:
                    return dumper(by_alias=True, exclude_none=True)
                except Exception:  # noqa: BLE001
                    try:
                        return dumper()
                    except Exception:  # noqa: BLE001
                        return {}
        return {}

    def render(self, value: Any) -> tuple[str, int]:
        """Redact + truncate an arbitrary value, as (preview, original size).

        Public because `mcpobs.http` needs exactly the same treatment for HTTP
        bodies, and a second copy of the redaction rules is how the two drift
        apart until one of them leaks.
        """
        return self._render(value)

    # -- internals ---------------------------------------------------------
    def _render(self, value: Any) -> tuple[str, int]:
        if value is None or value == "":
            return "", 0
        if not isinstance(value, str):
            cleaned = self._scrub(value) if self.redact else value
            try:
                # indent=2 deliberately: this is read by a human in a panel, not
                # parsed by a machine, and a one-line JSON-RPC envelope with
                # _meta inside it is unreadable.
                text = json.dumps(cleaned, ensure_ascii=False, default=str, indent=2)
            except (TypeError, ValueError):
                text = str(cleaned)
        else:
            text = self._scrub_text(value) if self.redact else value

        size = len(text)
        if size > self.max_chars:
            return text[: self.max_chars] + f"… (+{size - self.max_chars} chars)", size
        return text, size

    def _scrub(self, value: Any) -> Any:
        """Redact by key name, recursively, before serialising."""
        if isinstance(value, dict):
            return {
                k: REDACTED if self._is_sensitive_key(k) else self._scrub(v)
                for k, v in value.items()
            }
        if isinstance(value, list):
            # Non-text content blocks are summarised by TYPE, never dumped: a
            # base64 image is megabytes of noise telling an operator nothing.
            return [self._scrub(self._summarise_block(v)) for v in value]
        if isinstance(value, str):
            return self._scrub_text(value)
        return value

    def _scrub_text(self, text: str) -> str:
        for pattern in SENSITIVE_VALUES:
            text = pattern.sub(REDACTED, text)
        return text

    @staticmethod
    def _summarise_block(value: Any) -> Any:
        if isinstance(value, dict) and value.get("type") in ("image", "audio", "blob"):
            kept = {k: v for k, v in value.items() if k not in ("data", "blob")}
            kept["data"] = f"[{value.get('type')} omitted]"
            return kept
        return value

    @staticmethod
    def _is_sensitive_key(key: Any) -> bool:
        lowered = str(key).lower()
        if lowered in NEVER_REDACT:
            return False
        return any(marker in lowered for marker in SENSITIVE_KEYS)

File: mcpobs-js/src/payload.ts
/**
 * Payload capture and redaction. Off by default (`capturePayloads: false`),
 * mirroring the Python SDK's `mcpobs/payload.py` -- capturing arguments and
 * results is a deliberate opt-in, not a thing that happens because
 * instrumentation exists.
 *
 * Redacted BEFORE storage, not before display: a secret that reaches the
 * span cannot be recalled later. Same reasoning as the Python module's own
 * comment on this.
 */

const MAX_PREVIEW = 2_000;

const SECRET_KEY_PATTERN = /(api[_-]?key|token|password|secret|authorization|access[_-]?key)/i;

function redactValue(key: string, value: unknown): unknown {
  if (SECRET_KEY_PATTERN.test(key)) return "[redacted]";
  if (typeof value === "string") return redactString(value);
  if (value instanceof Error) return asDict(value);
  if (Array.isArray(value)) return value.map((item) => redactValue(key, item));
  if (value && typeof value === "object") return redactObject(value as Record<string, unknown>);
  return value;
}

/**
 * Best-effort for a model rather than the sealed wire form -- mirrors
 * mcpobs/payload.py's `_as_dict` docstring exactly, for the same reason.
 *
 * Error is the concrete case that motivated this: `message` and `stack` are
 * OWN but NON-ENUMERABLE properties, so `JSON.stringify(new Error("x"))`
 * silently produces `"{}"` -- measured directly, not assumed. A thrown
 * exception reaching payload capture (the tools/call exception path, or a
 * customer's own tool returning one) would have its actual text discarded
 * without this. `stack` is deliberately excluded, not just missed -- a stack
 * trace is exactly the kind of internal-structure/file-path detail this
 * project has been conservative about letting leave the process elsewhere
 * (docs/decisions.md's payload-capture boundaries make the same call).
 */
function asDict(err: Error): Record<string, unknown> {
  return { name: err.name, message: err.message };
}

// Recognizable secret FORMATS embedded in an otherwise plain string -- the
// only redaction an opaque string (an HTTP body, a URL) can get, since there
// is no key name to check. Matches mcpobs/payload.py's SENSITIVE_VALUES
// exactly, patterns and all: a raw request body with a `password` field does
// NOT get key-based redaction here, deliberately, same as Python (D56) --
// key-based redaction only applies to structured objects (tool args/results),
// where a real key name exists to check.
const SENSITIVE_VALUE_PATTERNS: readonly RegExp[] = [
  /\bBearer\s+[\w.~+/=-]{8,}/gi,
  /\bsk-[A-Za-z0-9]{16,}/g, // OpenAI-style
  /\bgh[pousr]_[A-Za-z0-9]{16,}/g, // GitHub
  /\bAKIA[0-9A-Z]{16}\b/g, // AWS access key id
  /\beyJ[\w-]{10,}\.[\w-]{10,}\./g, // JWT
  /([?&](?:api[_-]?key|token|access[_-]?key)=)[^&\s]+/gi,
];

function redactString(value: string): string {
  let out = value;
  for (const pattern of SENSITIVE_VALUE_PATTERNS) {
    out = out.replace(pattern, (match, group1?: string) => (group1 ? `${group1}[redacted]` : "[redacted]"));
  }
  return out;
}

function redactObject(obj: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(obj)) {
    out[key] = redactValue(key, value);
  }
  return out;
}

export interface CapturedPayload {
  preview: string;
  size: number;
}

export function capturePayload(value: unknown): CapturedPayload {
  let redacted: unknown;
  try {
    // Always through redactValue, never redactObject directly -- redactValue
    // already falls through to redactObject for plain objects on its own,
    // and going through it uniformly is what makes the Error special-case
    // apply at the TOP LEVEL too, not just when an Error is nested inside a
    // wrapper object (dispatch.ts's actual usage; a future direct call with
    // a bare Error must not regress the same bug for lack of this).
    redacted = redactValue("", value);
  } catch {
    return { preview: "", size: 0 };
  }
  const serialized = typeof redacted === "string" ? redacted : JSON.stringify(redacted);
  const text = typeof serialized === "string" ? serialized : "";
  return {
    preview: text.length > MAX_PREVIEW ? `${text.slice(0, MAX_PREVIEW)}...[truncated]` : text,
    size: text.length,
  };
}

export function capturePreview(value: unknown): string {
  return capturePayload(value).preview;
}

/**
 * ALLOW-listed HTTP headers, matching mcpobs/http.py's KEEP_HEADERS exactly
 * -- an allow-list, not a deny-list, on purpose. A deny-list is wrong by
 * default: it captures every header nobody thought of, and the one nobody
 * thought of is where the next bespoke `x-acme-session` credential lives.
 * `authorization` and `cookie` are never READ at all, not merely redacted.
 */
const KEEP_HEADERS = new Set([
  "content-type",
  "content-length",
  "accept",
  "accept-encoding",
  "user-agent",
  "x-request-id",
  "traceparent",
  "retry-after",
  "cache-control",
  "location",
  "server",
]);

/**
 * Renders only the allow-listed headers, through the same redaction/preview
 * pipeline as everything else -- one implementation, so header capture and
 * MCP payload capture cannot drift into different redaction rules (same
 * reasoning as http.py's module docstring).
 */
export function captureHeaders(headers: Record<string, unknown> | undefined | null): string {
  if (!headers) return "";
  const kept: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(headers)) {
    if (KEEP_HEADERS.has(key.toLowerCase())) kept[key] = value;
  }
  if (Object.keys(kept).length === 0) return "";
  return capturePreview(kept);
}
